### PR TITLE
VGG, EfficientNet, and Resnet model tests

### PR DESCRIPTION
**Pull request by Sara-Eftekhar-Azam**
I added few different versions of ResNet model.

### CODE REVIEW

**Comment by Bastiaan27** [comment]
Does this fully work and was this properly tested?

**Review by Bastiaan27 — COMMENTED** [review]
Does this fully work and was this properly tested?

File: code/modelconcepts/EfficientNet/efficient_net_b1_train.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+
+dataset = RooftileDataset(root_dir=pathToDateset, folders=folders_to_use, transform=transform)
+
+# Split dataset into train and test
+train_size = int(train_ratio * len(dataset))
+test_size = len(dataset) - train_size
+train_dataset, test_dataset = torch.utils.data.random_split(dataset, [train_size, test_size])
+
+train_loader = DataLoader(train_dataset, batch_size=32, shuffle=True)
+test_loader = DataLoader(test_dataset, batch_size=32, shuffle=False)
```
Comment (Bastiaan27) [inline code/modelconcepts/EfficientNet/efficient_net_b1_train.py:107]: Read the readme.md to look how to split the dataset into test and training, please don't split the dataset here in the file

**Review by Bastiaan27 — CHANGES_REQUESTED** [review]

File: code/modelconcepts/EfficientNet/efficient_net_b1_train.py (hunk 1 above)
Comment (Sara-Eftekhar-Azam) [reply code/modelconcepts/EfficientNet/efficient_net_b1_train.py:107]: I looked at readme.md but it was not completely clear to me. Do you mean that I should use train and test folders? This was a first version. We can definitely improve it.

Comment (Sara-Eftekhar-Azam) [reply code/modelconcepts/EfficientNet/efficient_net_b1_train.py:107]: I suggest that we merge this so far and then I will create a new branch to make these changes. I plan to add a confusion matrix too in the next version.

Comment (Phaneron97) [reply code/modelconcepts/EfficientNet/efficient_net_b1_train.py:107]: No, we won't merge this as there is already code written that stratified splits everything into train and test. Please preprocess your dataset on your own local machine with the help of the [readme](https://github.com/Phaneron97/BDA_proj_rooftiles/blob/main/code/data_analysis/readme.md) and follow the steps carefully. It's not difficult, just run the files in order one by one. After preprocessing, edit your model so that it expect an already preprocessed dataset. If we do this for all models we can keep it clean.

**Review by Phaneron97 — CHANGES_REQUESTED** [review]
Don't push your latex compile files to github, these are yours only and can differ from others. It also clutters the repo. remove your cached additions locally with `git rm -r --cached <latex compile files>` and push again.

File: code/modelconcepts/EfficientNet/efficient_net_b1_train.py (hunk 1 above)
Comment (Sara-Eftekhar-Azam) [reply code/modelconcepts/EfficientNet/efficient_net_b1_train.py:107]: OK. I will change my code.

**Comment by Sara-Eftekhar-Azam** [comment]
> Don't push your latex compile files to github, these are yours only and can differ from others. It also clutters the repo. remove your cached additions locally with `git rm -r --cached <latex compile files>` and push again.

I tried to do this but I am not sure if I am doing this right. I will create a new branch from development and make changes there and apply feedbacks and create a new pull request

**Comment by Sara-Eftekhar-Azam** [comment]
I changed my code. The code now uses the train and test folders created by running the last sprint methods. 
I also changed my code to make it have more methods and be more organized. Also, instead of one file for each model (for example 8 files for efficientnet b0 to b7) I created one file for each model that gets the model version as a parameter. 
I also added plotting confusion matrix at the end.
I ran my code on Google Colab and tested it with 20 images for each class. It got to high test accuracies (even more than 90%) 
I think we just have to run the code for different versions and add them to our report. Annemarie said that running each model for a little time or part of the data is enough to help us choose between different models an then we can run our selected model on all of the data and longer.

**Comment by Phaneron97** [comment]
> Annemarie said that running each model for a little time or part of the data is enough to help us choose between different models an then we can run our selected model on all of the data and longer.

This is where hyperparameter tuning comes in. According to Marya one could use one Keras Tuners to find the best setting which can yield the highest accuracy. We'll have to look into that once we have some models up and running.

**Comment by Sara-Eftekhar-Azam** [comment]
> > Annemarie said that running each model for a little time or part of the data is enough to help us choose between different models an then we can run our selected model on all of the data and longer.
> 
> This is where hyperparameter tuning comes in. According to Marya one could use Keras Tuner to find the best setting which can yield the highest accuracy. We'll have to look into that once we have some models up and running.

That is a good suggestion. I have used pytorch library (and not Keras). Although there should be similar things in pytorch as well for hyperparameter tuning. I think it needs a good computer and some time to find best parameters. I tried some and found some that were good in my opinion. The code works now. So, feel free to change it and add hyperparameter tuning to it. I will try to test few more cases and also start writing the report.

**Comment by Phaneron97** [comment]
> > > Annemarie said that running each model for a little time or part of the data is enough to help us choose between different models an then we can run our selected model on all of the data and longer.
> > 
> > 
> > This is where hyperparameter tuning comes in. According to Marya one could use Keras Tuner to find the best setting which can yield the highest accuracy. We'll have to look into that once we have some models up and running.
> 
> That is a good suggestion. I have used pytorch library (and not Keras). Although there should be similar things in pytorch as well for hyperparameter tuning. I think it needs a good computer and some time to find best parameters. I tried some and found some that were good in my opinion. The code works now. So, feel free to change it and add hyperparameter tuning to it. I will try to test few more cases and also start writing the report.

Yeah and even cooler, you can use I think it's called Hyperband to stop spending more time learning a model once the tuner notices there are no improvements made. This can save a lot of time.
